hub/metrics.go

```diff
@@ -0,0 +1,80 @@
+package hub
+
+import (
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Metrics store Hub collected metrics.
+type Metrics struct {
+	subscribersTotal *prometheus.CounterVec
+	subscribers      *prometheus.GaugeVec
+	updatesTotal     *prometheus.CounterVec
+}
+
+// NewMetrics creates a Prometheus metrics collector.
+func NewMetrics() *Metrics {
+	return &Metrics{
+		subscribersTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "mercure_subcribers_total",
+				Help: "Total number of handled subsribers",
+			},
+			[]string{"topic"},
+		),
+		subscribers: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "mercure_subcribers",
+				Help: "The current number of running subscribers",
+			},
+			[]string{"topic"},
+		),
+		updatesTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "mercure_updates_total",
+				Help: "Total number of handled updates",
+			},
+			[]string{"topic"},
+		),
+	}
+}
+
+// Register configure the Prometheus registry with all collected metrics.
+func (m *Metrics) Register(r *mux.Router) {
+	registry := prometheus.NewRegistry()
+
+	// Metrics about the Hub
+	registry.MustRegister(m.subscribers)
+	registry.MustRegister(m.subscribersTotal)
+	registry.MustRegister(m.updatesTotal)
+
+	// Go-specific metrics about the process (GC stats, goroutines, etc.).
+	registry.MustRegister(prometheus.NewGoCollector())
+	// Go-unrelated process metrics (memory usage, file descriptors, etc.).
+	registry.MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+
+	r.Handle("/metrics", promhttp.HandlerFor(registry, promhttp.HandlerOpts{})).Methods("GET")
+}
+
+// NewSubscriber collect metrics about new subscriber event.
+func (m *Metrics) NewSubscriber(topics []string) {
```

Comment:
You should also expose the targets and the IP. Here are the data we currently log: https://github.com/dunglas/mercure/blob/master/hub/log.go#L11-L31

Reply:
I'm not sure that IP information is relevant in metrics (metrics are not logs).

We could add targets as a label. The output result will me something like:

```text
mercure_subcribers{topic="xxx", target="yyy"}
mercure_subcribers_total{topic="xxx", target="yyy"}
mercure_updates_total{topic="xxx", target="yyy"}
```

But with this, as each label creates a unique combination of time series, it will increase the amount of data.

Reply:
Ok let's keep it as is then!